src/core/control/xojfile/LoadHandler.cpp

```diff
@@ -394,528 +248,365 @@ void LoadHandler::parseBgPixmap() {
-            return;
-        }
+void LoadHandler::parseXml() {
+    xoj_assert(this->doc);
+    xoj_assert(this->xmlContentStream);
 
-        gboolean writeSuccessful =
-                g_output_stream_write_all(outputStream, data, static_cast<gsize>(read), nullptr, nullptr, nullptr);
-        if (!writeSuccessful) {
-            g_free(data);
-            zip_fclose(attachmentFile);
-            error("%s", FC(_F("Could not open attachment: {1}. Error message: Could not write file") % filename));
-            return;
-        }
+    XmlParser parser(*this->xmlContentStream, this);
+    auto res = parser.parse();
 
-        readBytes += static_cast<zip_uint64_t>(read);
+    if (res != 0) {
+        logError("File is not finished after end of document");
     }
-    g_free(data);
-    zip_fclose(attachmentFile);
-
-    g_hash_table_insert(this->audioFiles, g_strdup(filename), g_file_get_path(tmpFile.get()));
-}
-
-void LoadHandler::parserStartElement(GMarkupParseContext* context, const gchar* elementName,
-                                     const gchar** attributeNames, const gchar** attributeValues, gpointer userdata,
-                                     GError** error) {
-    auto* handler = static_cast<LoadHandler*>(userdata);
-    // Return on error
-    if (*error) {
-        return;
+    if (!this->parsingComplete) {
+        throw std::runtime_error(_("Document is not complete (maybe the end is cut off?)"));
     }
-    handler->attributeNames = attributeNames;
-    handler->attributeValues = attributeValues;
-    handler->elementName = elementName;
-
-    if (handler->pos == PARSER_POS_NOT_STARTED) {
-        handler->parseStart();
-    } else if (handler->pos == PARSER_POS_STARTED) {
-        handler->parseContents();
-    } else if (handler->pos == PARSER_POS_IN_PAGE) {
-        handler->parsePage();
-    } else if (handler->pos == PARSER_POS_IN_LAYER) {
-        handler->parseLayer();
-    } else if (handler->pos == PARSER_POS_IN_IMAGE || handler->pos == PARSER_POS_IN_TEXIMAGE) {
-        // Handle the attachment node within the appropriate nodes
-        if (!strcmp(elementName, "attachment")) {
-            handler->parseAttachment();
-        }
+    if (this->parsingComplete && this->doc->getPageCount() == 0) {
+        throw std::runtime_error(_("Document is corrupted (no pages found in file)"));
     }
 
-    handler->attributeNames = nullptr;
-    handler->attributeValues = nullptr;
-    handler->elementName = nullptr;
+    doc->setCreateBackupOnSave(true);
 }
 
-void LoadHandler::parserEndElement(GMarkupParseContext* context, const gchar* elementName, gpointer userdata,
-                                   GError** error) {
-    // Return on error
-    if (*error) {
-        return;
+
+auto LoadHandler::loadDocument(fs::path const& filepath) -> std::unique_ptr<Document> {
+    initAttributes();
+    this->doc = std::make_unique<Document>(&dHanlder);
+
+    try {
+        openFile(filepath);
+        parseXml();
```

Comment:
So the idea here is that if we load the document into a local pointer, it is automatically destroyed if an exception occurs, right? That makes access to the Document by all the other member functions more complex. Is it really worth the trouble? 
It doesn't even really matter if we do not reset the pointer correctly upon failure: it would only loiter around until another document is loaded or the LoadHandler object itself is destroyed. Eventually, we could also manually reset all other building blocks like `this->page` or `this->text` in the catch-clause, since they would also be in an undefined and useless state in case of a failure. 

PS: lost my laptop 😬, new changes will have to wait a little bit

Reply:
That is the idea, yes. The thing is, that storing variables in a class for sequential use has always a code smell: it increases coupling and it makes it harder to track the data flow. The best is, to use local static functions, which only have access to the data they need.

Maybe just throw out of the loadhandler and catch it there.

Reply:
The sequential access pattern is a consequence of the current class architecture: the parser pushes information it reads sequentially from the file. I think it's a trade off for separating away XmlParser from LoadHandler. 

But this made me question if it wouldn't make sense to split the LoadHandler up further, putting the document building callback stuff into a separate class. The LoadHandler would only worry about file handling, setting up the document builder, and finishing up at the end. The XmlParser would do all its calls to the document builder. The document builder would still need to go back to the LoadHandler for attachments and the like. Does that sound interesting or would that only cause more complexity?

And yes, throwing out of the load handler sounds like a good idea.

Reply:
> And yes, throwing out of the load handler sounds like a good idea.

For now probably the best, this will make the LoadHandler way more readable and safer.
A variant, using an own implemented version of `std::expected` is also nice, but before c++ isn't adding something like try or Rusts `?` operator for `std::expected` this is cumbersome or requires defines like `#define CHECK_EX(e) if(!e) return {e.error()};`. 

> The sequential access pattern is a consequence of the current class architecture: the parser pushes information it reads sequentially from the file. I think it's a trade off for separating away XmlParser from LoadHandler.
> 
> But this made me question if it wouldn't make sense to split the LoadHandler up further, putting the document building callback stuff into a separate class. The LoadHandler would only worry about file handling, setting up the document builder, and finishing up at the end. The XmlParser would do all its calls to the document builder. The document builder would still need to go back to the LoadHandler for attachments and the like. Does that sound interesting or would that only cause more complexity?


Probably, refactoring this, may be desirable, but that's something we should lay out on paper with many iterations. It's also questionable, if we need a class hierarchy here, or if free functions, returning a `expected<unique_ptr<Document>, Exception>` are sufficient.
It may also be good, to have several FileHandlers, for different formats.
Also, when doing this, the loading of a xopp file should be a completely self-contained static library, with an implementation to load xournal Documents.

So we can define something like:
```
struct XoppDocumentTraits {
    using DocType = Document;
    using ReturnType = std::unique_ptr<Document>;

    constexpr static auto layerHandler = ...
    constexpr static auto strokeHandler = ...
    //...
};

template<class Traits>
class XoppIOHandler {
    auto load(LoadOptions<Traits>) -> std::expected<ReturnType, std::exception>;
    auto safe(SafeOptions<Traits>, DocType const&) -> std::exception
}

constexpr auto singletonXoppDocIO = XoppIOHandler<XoppDocumentTraits>;
```